### DISCUSSION TITLE
Clean "unplug" of virtual disk drive

### DISCUSSION THREAD

**Discussion by jernst** [utmapp/UTM#5866] [UTM Help (macOS)]
How do I cleanly remove a virtual disk from a VM? I.e. what's the equivalent of unplugging a SATA disk when the machine is off?

If there is no such functionality, please add in UI and in the API. Presumably it would:
* delete the .qcow2 file
* edit the plist file

For comparison, this is what VirtualBox does: https://www.virtualbox.org/manual/ch05.html

**Comment by osy** [comment]
What are you trying to do? There is no such thing as a "clean unplug". When the VM is powered off, you can do whatever you want.

**Comment by jernst** [comment]
I can think of many use cases, here are two:

* I want to create a new Linux VM from scratch. So I run a "bootstrap" Linux VM, attach a separate blank virtual disk, and from the boostrap VM I do thinks like fdisk, mkfs, and all the things I want to do to create my VM on the extra disk. When I'm done, I need to unattach that separate disk, and turn it into the root disk of my new VM.

* I have a Linux VM that for some reason won't boot. I run a separate VM, attach the root disk of the broken VM, and then mkfs and whatever else I need to do to get it back to boot. Then I unattach the disk, re-attach to the original VM.

This process is possible today, but very clumsy and very error-prone (as I experienced first-hand myself): basically, I need to quit everything, move files around and edit the plist file consistently with the way I moved the files. And better not forget which disk had which UUID.

Other virtualization apps, like VirtualBox or Parallels, have little buttons to "unattach" and "reattach" disks just like UTM has one for "add disk". Also, a corresponding API would be nice, see for example https://www.virtualbox.org/manual/ch08.html#vboxmanage-storagectl
